[Darren Teo] iP

darrtzj opened the pull request:
# Duke 

> "Just keep swimming." - **Dory**, _Finding Nemo_ ([source](http://www.planetofsuccess.com/blog/2019/movie-quotes/#funny))

Duke frees your mind so that you can focus on the things that matter. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to user

All you need to do is,

1.  download it from [here](https://github.com/darrtzj/ip/releases/tag/A-Jar).
2.  double-click it.
3.  add your tasks.
4.  let it manage your tasks for you 👍 👍 👍 

And it is **FREE!**

Features:

- [x] Keep track of tasks
- [ ] Reminder (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
        new Duke(filePath).run();
    }
```

darrtzj committed:
Level 1. Greet, Echo, Exit

darrtzj committed:
added ArrayList to store text

darrtzj committed:
added list functionality

darrtzj committed:
added Task class

darrtzj committed:
modified Duke.java to use Task class

darrtzj committed:
modified to fit level 3 test case given

darrtzj committed:
Added tasks sub classes

darrtzj committed:
abstracted overlapping code in todo, deadline and event to the start of the loop

darrtzj committed:
updated test cases

darrtzj committed:
added error handling for incorrect inputs

darrtzj committed:
added delete function

darrtzj committed:
updated test case

darrtzj committed:
Add Save function, with save path creating a data directory
and data saved in duke.txt file.

darrtzj committed:
Change time in deadline and event from String to LocalDate

darrtzj committed:
Merge branch 'branch-Level-7'

darrtzj committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

darrtzj committed:
Change parts that take time as String to LocalDate

darrtzj committed:
DukeException class added to handle exceptions
Storage class added to handle loading and saving
TaskList class added to abstract the tasklist

darrtzj committed:
updated test cases

darrtzj committed:
Added Ui class

darrtzj committed:
Added Command class and multiple sub classes for ui commands
Added Parser to read user command

darrtzj committed:
updated test cases

darrtzj committed:
Organize classes into packages

darrtzj committed:
Set up gradle

darrtzj committed:
Added JUnit test for Todo and Deadline

darrtzj committed:
Modified build.gradle for Jar file creation

darrtzj committed:
Add JavaDocs for all classes

darrtzj committed:
Tweak code to follow coding standards

darrtzj committed:
Fix bug at Parser

darrtzj committed:
Add find function to duke

darrtzj committed:
Add override tag to command classes

darrtzj committed:
Merge branch 'branch-Level-9'

darrtzj committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/ExitCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/command/MarkCommand.java

darrtzj committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java
#	src/main/java/duke/command/AddCommand.java
#	src/main/java/duke/command/Command.java
#	src/main/java/duke/command/DeleteCommand.java
#	src/main/java/duke/command/ExitCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/command/MarkCommand.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/Event.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/Todo.java

darrtzj committed:
Add checkstyle test and tweak code to follow coding standard

kxrt commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,59 @@
+    private Ui ui;
+
+    /**
+     * Initialize duke
```
The [coding standard](https://se-education.org/guides/conventions/java/intermediate.html#comments) specifies Javadoc comments need proper punctuation. Perhaps it would be better to reformat all your Javadoc comments to follow this.

kxrt commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,132 @@
+    /**
+     * Displays all task in task list
+     *
+     * @param tasks Task list
```
Perhaps it would be better to add the error thrown into the Javadoc comments in general across all your code, as the [Oracle Javadoc standard](https://www.oracle.com/sg/technical-resources/articles/java/javadoc-tool.html#throwstag) specifies all checked exceptions must be declared.

kxrt commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,59 @@
+ */
+public class Duke {
+
+    private static final String filePath = "./data/duke.txt";
```
Capitalising the variable name here would be in line with the coding standard for constants.

kxrt commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,163 @@
+        String taskDescription = taskDetails[2];
+        boolean isDone;
+        isDone = taskDetails[1].equals("1");
+        if (taskDetails.length == 4) {
```
I think it would be better to add a comment specifying the meaning behind hardcoded values (i.e. 4).

kxrt commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,132 @@
+ */
+public class Ui {
+
+    private static final String GREETINGS = "Hello! I'm Duke\n" + "What can I do for you?";
```
Perhaps appending "_MESSAGE" here would be more in line with the coding standard, which specifies associated constants should have a common prefix. Maybe shifting the "MESSAGE" keyword itself to be a prefix instead of a suffix would be better too.

kxrt commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,39 @@
+     * @param ui User interface of programme
+     * @param storage Storage of programme
+     * @throws DukeException If error saving file
+     */
```
Maybe using the `@inheritDoc` property mentioned in the coding standard across classes extended from `Command.java` would be better in saving lines of code, with added modifications for changed behaviour.

kxrt commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,147 @@
+ */
+public class Parser {
+
+    private static final String EXIT_COMMAND = "bye";
```
Perhaps shifting the "COMMAND" suffix to be a prefix would be more in line with the coding standard.

kxrt reviewed (commented):
Code formatting is generally good to go and very clean, some minor Javadoc issues that can be fixed pretty quickly! Good luck for the rest of iP mate :)

darrtzj committed:
Tweak Javadoc comments and variable name to align with
coding standard

darrtzj committed:
Add base code for GUI

darrtzj committed:
Change duke interface from text ui to a gui

darrtzj committed:
Add greeting
Add exit gui on user command

darrtzj committed:
Update javadocs

darrtzj committed:
Update javadoc

darrtzj committed:
Update code to follow coding standard

darrtzj committed:
Merge branch 'branch-Level-10'

darrtzj committed:
Update main class in Gradle

darrtzj committed:
Add assertions

No assertions is being used to verify program state.

If program is not in assumed state due to bugs, running the program
may cause problems. The sooner the execution is stopped, the safer
it is.

Assertions are added to TaskList to verify date format and task
format of data loaded from save file.

darrtzj committed:
Task classes: reduce magic strings

Task classes contain many magic strings.

Magic strings do not provide information on what the string does.

Magic strings have been replaced with a meaningful variable names.

This allows future developer to identify the purpose of the string.

darrtzj committed:
Remove magic strings

darrtzj committed:
Remove magic strings and magic numbers

darrtzj committed:
Extracted magic string in MainWindow

darrtzj committed:
Merge pull request #5 from darrtzj/branch-A-Codequality

Branch a codequality

darrtzj committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/task/Task.java

darrtzj committed:
Merge pull request #6 from darrtzj/branch-A-Assertions

Branch a assertions

darrtzj committed:
Add function to add and delete alias for commands

darrtzj committed:
Merge branch 'branch-C-FriendlierSyntax'

darrtzj committed:
Add Ui.png

darrtzj committed:
Update README.md with user guide

darrtzj committed:
Add bye command into user guide
